Docs: update custom domains docs (#9266)
readthedocs/readthedocs.org: docs/user/canonical-urls.rst

Canonical URLs
==============

> Canonical URLs allow you to have consistent page URLs across all your domains.

Reviewer: I think this needs more explanation. What does it mean "consistent page URLs across all your domains" for a user?

This is mainly useful for search engines,
so that they can send users to the correct version and domain of your documentation.

The canonical URL is generated based on:

* The default version of your project (usually "latest" or "stable").
* The canonical :doc:`custom domain </custom-domains>` if you have one,

>   otherwise the default :ref:`subdomain <hosting:subdomain support>` will be used.

Reviewer: This mentions how the canonical URL is generated but does not tell me what's the ending result in each case. I'm not sure to understand these bullets.

.. contents::
    :local:

Example
-------

Fabric hosts their docs on Read the Docs,
but they use their own domain ``http://docs.fabfile.org``.
This means that Google will index both ``http://fabric-docs.readthedocs.io`` and
``http://docs.fabfile.org`` for their documentation.

Fabric will want to set ``http://docs.fabfile.org`` as their canonical domain,
this means that when Google indexes ``http://fabric-docs.readthedocs.io``,

> it will know that it should really point at ``http://docs.fabfile.org``.

Reviewer: I think it will be useful to merge this example section with the previous bullets and generalize the example without using Fabric but just `example.com` or similar.

Implementation
--------------

If you are using :doc:`Sphinx </intro/getting-started-with-sphinx>`,
Read the Docs will set the value of the html_baseurl_ setting (if isn't already set) to your canonical domain.

.. _html_baseurl: https://www.sphinx-doc.org/page/usage/configuration.html#confval-html_baseurl

For :doc:`MkDocs </intro/getting-started-with-mkdocs>` this isn't done automatically,
but you can use the site_url_ setting.

> .. _site_url: https://www.mkdocs.org/user-guide/configuration/#site_url

Reviewer: These should probably be subsections with instead just 2 paragraph. Also, it could mention that this does not work with `build.commands` maybe...

If you look at the source code for the documentation built after you set your canonical URL,
you should see a bit of HTML like this:

.. code-block:: html

   <link rel="canonical" href="http://docs.fabfile.org/en/2.4/" />

> .. note::
> 
>    If you change your default version or canonical domain,
>    you'll need to re-build all your versions in order to update their
>    canonical URL to the new one.

Reviewer: ```suggestion
.. warning::

   If you change your default version or canonical domain,
   you'll need to re-build all your versions in order to update their
   canonical URL to the new one.
```

Looks important enough to make it more prominent.

Canonical URLs
==============

`Canonical URLs`_ allow you to specify the preferred version of a web page to prevent duplicated content.
They are mainly used by search engines to link users to the correct
version and domain of your documentation.

> .. _Canonical URLs: https://datatracker.ietf.org/doc/html/rfc6596

Reviewer 2: We definitely don't want to link to this. Perhaps https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls?

.. contents:: Contents
    :local:

Example
-------

The canonical URL is generated based on:

* The default version of your project (usually "latest" or "stable").
* The canonical :doc:`custom domain </custom-domains>` if you have one,
  otherwise the default :ref:`subdomain <hosting:subdomain support>` will be used.

> For example, if you have a project named ``docs``
> with a custom domain ``https://docs.example.com``.
> A search engine like Google will index both domains, ``http://docs.readthedocs.io`` and
> ``https://docs.example.com``.

Reviewer 2: ```suggestion
For example, if you have a project named ``example-docs``
with a custom domain ``https://docs.example.com``.
A search engine like Google will index both domains, ``http://example-docs.readthedocs.io`` and
``https://docs.example.com``.
```

But you'll want to use ``https://docs.example.com`` as your canonical domain,
this means that when Google indexes a page like ``https://docs.readthedocs.io/en/9.0/``,
it will know that it should really point at ``https://docs.example.com/en/latest/``.

> 

Reviewer 2: This feels really confusing to a user... I feel like we aren't explaining why this is useful here.

Implementation
--------------

If you are using :doc:`Sphinx </intro/getting-started-with-sphinx>`,
Read the Docs will set the value of the html_baseurl_ setting (if isn't already set) to your canonical domain.

.. _html_baseurl: https://www.sphinx-doc.org/page/usage/configuration.html#confval-html_baseurl

For :doc:`MkDocs </intro/getting-started-with-mkdocs>` this isn't done automatically,
but you can use the site_url_ setting.

.. _site_url: https://www.mkdocs.org/user-guide/configuration/#site_url

If you look at the source code for the documentation built after you set your canonical URL,
you should see a bit of HTML like this:

.. code-block:: html

   <link rel="canonical" href="http://docs.fabfile.org/en/2.4/" />

.. warning::

   If you change your default version or canonical domain,
   you'll need to re-build all your versions in order to update their

>    canonical URL to the new one.

Reviewer 2: This is bad UX :(

Author: This warning applies for both, mkdocs and sphinx, we should move it at the parent section maybe

Canonical URLs
==============

A `Canonical URL`_
allows you to specify the preferred version of a web page to prevent duplicated content.
They are mainly used by search engines to link users to the correct
version and domain of your documentation.

If canonical URL's aren't used,
it's easy for outdated documentation to be the top search result for various pages in your documentation.

> This is not a perfect solution for this problem,
> but generally people finding outdated documentation is a big problem,
> and this is one of the suggested ways to solve it from search engines.

Reviewer 3: This feels redundant, and/or isn't adding much for the reader. Is there a specific reason worth mentioning why it isn't a great solution perhaps? If not, maybe worth skipping this sentence

```suggestion
```

Reviewer 3: Agreed, but the sentence before describes the problem better. This sentences mostly distills down to "this is one imperfect solution to the problem we just outlined", and so sounds redundant, and isn't confident.

I'm suggesting either expanding on the problem/solution or dropping this sentence. Something like this expands on both:

```suggestion
Linking to outdated documentation is confusing for your readers,
but this can be avoided by telling search engines exactly where to find your documentation.
```

.. _Canonical URL: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls

How Read the Docs generates canonical URLs
------------------------------------------

The canonical URL takes into account:

* The default version of your project (usually "latest" or "stable").
* The canonical :doc:`custom domain </custom-domains>` if you have one,
  otherwise the default :ref:`subdomain <hosting:subdomain support>` will be used.

> For example, if you have a project named ``example-docs``
> with a custom domain ``https://docs.example.com``.
> A search engine like Google will index both domains, ``https://example-docs.readthedocs.io`` and
> ``https://docs.example.com``.

Reviewer 3: This was a little fragmenty, but readers might not make the connection between `example-docs` and it's hosted domain `example-docs.readthedocs.io`

```suggestion
For example, if you have a project named ``example-docs``
with a custom domain ``https://docs.example.com``,
then your documentation will be served at ``https://example-docs.readthedocs.io`` and ``https://docs.example.com``.
Without specifying a canonical URL, a search engine like Google will index both domains.
```

> You'll want to use ``https://docs.example.com`` as your canonical domain,

Reviewer 3: ```suggestion
You'll want to use ``https://docs.example.com`` as your canonical domain.
```

> this means that when Google indexes a page like ``https://example-docs.readthedocs.io/en/latest/``,

Reviewer 3: ```suggestion
This means that when Google indexes a page like ``https://example-docs.readthedocs.io/en/latest/``,
```

Reviewer 2: @agjohnson You can do multi-line suggestions in cases like this to make review easier.

Canonical URLs
==============

> A `Canonical URL`_

Reviewer 3: ```suggestion
A `canonical URL`_
```

but generally people finding outdated documentation is a big problem,
and this is one of the suggested ways to solve it from search engines.

> .. _Canonical URL: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls

Reviewer 3: ```suggestion
.. _canonical URL: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls
```
